docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1432 @@
+# KEP-2170: Kubeflow Training V2 API
+
+## Authors
+
+- Andrey Velichkevich - [@andreyvelich](https://github.com/andreyvelich)
+- Yuki Iwai - [@tenzen-y](https://github.com/tenzen-y)
+
+Creation date: 2024-07-16
+
+Google doc: https://bit.ly/3WzjTlw
+
+## Overview
+
+This document discusses the new Kubeflow Training V2 API.
+
+When we built the
+[Kubeflow Training Operator a couple of years ago](https://docs.google.com/document/d/1x1JPDQfDMIbnoQRftDH1IzGU0qvHGSU4W6Jl4rJLPhI/edit?usp=sharing),
+Kubernetes lacked better features to support distributed machine learning (ML) training, such as
+SuccessPolicy and RestartPolicy (FailurePolicy). Recently, the Kubernetes community launched the
+working group Batch, and then the working group actively worked on evolving the batch/v1 `Job` API
+and built [a new Kubernetes SIGs project: `JobSet`](https://github.com/kubernetes-sigs/jobset) to
+manage groups of `Jobs`.
+
+This document consolidates efforts for the Cloud Native ML Training between Kubeflow and Kubernetes
+communities.
+
+## Motivation
+
+We often implement features similar to batch/v1 `Job`, such as “suspend”, on the Training Operator
+side since the Training Operator creates blocks of plain Pod and Service for each rank once
+Kubeflow Jobs are created. However, if we continue taking the same approach to use lowest level
+abstractions that introduce redundancy, the maintenance costs will continue to increase.
+
+Replacing repetitive infrastructure layers with `JobSet` would help to avoid redundancy and reduce
+developer toil.
+
+Additionally, introducing `JobSet` as an infrastructure layer would allow us to introduce batch
+workload features such as
+[the PodFailurePolicy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy)
+and [the PodDisruptionCondition](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
+easily.
+
+Please also see the [Kubernetes JobSet and Kubeflow Training Operator collaboration document](https://docs.google.com/document/d/1C2ev7yRbnMTlQWbQCfX7BCCHcLIAGW2MP9f7YeKl2Ck/edit?usp=sharing).
```

Comment:
This is great! And it's really helpful context. I briefly reviewed the JobSet docs / concepts and it's intuitive why we would want to move fault tolerance into a separate project. 

In previous roles, along with having to deal with pod/job failures, I've also had to spend a lot of time on dealing with capacity constraints/issues especially among different teams. Some folks at IBM recently shared with me work they've done and I'm curious to see where capacity management would be best fit for large-scale model training? 

Given the separation of pod failures into JobSet, I don't think it makes sense for capacity management to go in the Training Operator but would a separate project make more sense then?

Would love any thoughts here.

Reply:
> I'm curious to see where capacity management would be best fit for large-scale model training?

I think, we should discuss it with Batch WG (cc @tenzen-y @alculquicondor @ahg-g @kannon92 @danielvegamyhre). Especially the use-cases and what kind of capacity management we are talking about.

Reply:
Yes please. Defining "capacity issues" would be helpful but we should discuss outside of this KEP.

Reply:
Sounds good! I'll add it to the agenda.